datafusion/common/src/tree_node.rs

```diff
@@ -289,15 +492,80 @@ pub enum RewriteRecursion {
     Skip,
 }
 
-/// Controls how the [`TreeNode`] recursion should proceed for [`TreeNode::visit`].
+/// Controls how the [`TreeNode`] recursion should proceed for [`TreeNode::visit_down()`] and
+/// [`TreeNode::visit()`].
 #[derive(Debug)]
-pub enum VisitRecursion {
-    /// Continue the visit to this node tree.
+pub enum TreeNodeRecursion {
+    /// Continue the visit to the next node.
     Continue,
-    /// Keep recursive but skip applying op on the children
-    Skip,
-    /// Stop the visit to this node tree.
+
+    /// Prune the current subtree.
+    /// If a preorder visit of a tree node returns [`TreeNodeRecursion::Prune`] then inner
+    /// children and children will not be visited and postorder visit of the node will not
+    /// be invoked.
+    Prune,
+
+    /// Stop recursion on current tree.
+    /// If recursion runs on an inner tree then returning [`TreeNodeRecursion::Stop`] doesn't
+    /// stop recursion on the outer tree.
     Stop,
+
+    /// Stop recursion on all (including outer) trees.
+    StopAll,
+}
+
+impl TreeNodeRecursion {
+    pub fn and_then_on_continue<F>(self, f: F) -> Result<TreeNodeRecursion>
```

Comment:
These are neat helpers, it would be useful to document their intended usecases if possible